In [1]:
import json
import gzip
import jsonlines
import pandas as pd
import json
from pandas.io.json import json_normalize
import re

## 1. Converting Json Lines file to dataframe    
- The json-lines file is a list of json objects I read that into a stream and convert that stream into list of json objects    
- Then I take this list and convert to the dataframe by using json_normalize() which gives me the dataframe output as shown below

In [2]:
with gzip.GzipFile("ida_wrangling_exercise_data.2017-02-13.jsonl.gz", "rb") as f:
    stream = f.read().decode('utf-8')

In [3]:
data = []

# convert list of string to list of jsons 
for str in stream.split('\n'):
    if str!='': #is str is not empty in list
        data.append(json.loads(str))    

In [4]:
df = json_normalize(data)

In [5]:
df.head(10)

address     address.city  \
0                                                NaN        Hoodburgh   
1   20722 Coleman Villages\nEast Rose, SC 71064-5894              NaN   
2                                                NaN        New Julie   
3          0932 Gomez Drives\nLeefort, MD 46879-3166              NaN   
4                                                NaN  East Sharonstad   
5                                                NaN     Thompsonview   
6  9524 Danielle Burg Apt. 849\nNorth Vickistad, ...              NaN   
7         407 Patrick Shoal\nWilliamsshire, VI 49526              NaN   
8                                                NaN        Chrisview   
9                                                NaN              NaN   

  address.state               address.street address.zip         dob  \
0            RI    86314 David Pass Apt. 211       83973  1971-06-30   
1           NaN                          NaN         NaN  1965-09-09   
2            UT            6676 Young Square       73125  1993-04-12   
3           NaN                          NaN         NaN  1977-04-14   
4            ME              158 Smith Vista       42483  1970-03-19   
5            AR             461 Knapp Unions       02398  1989-12-23   
6           NaN                          NaN         NaN  1983-05-09   
7           NaN                          NaN         NaN  1989-10-05   
8            CT  3697 Mills Estates Apt. 499  70997-3753  1987-06-24   
9           NaN                          NaN         NaN  1980-03-20   

                                  email                                id  \
0                     opark@hotmail.com  01d68a4c598a45559c06f4df0b3d82cb   
1                  sperez@armstrong.com  876ff718291d4397bb1e0477ceee6ad9   
2                      uortiz@gmail.com  81753097bf7e4e2085982f422bdb9cda   
3                 palmerdiane@yahoo.com  2c2f7154b80f40ca80d08c5adc54ea45   
4                nancymaxwell@gmail.com  4f5263f339694d068e17ee7fdbb852b8   
5                 peterkhan@serrano.net  e615757bb70e4c9e8ec532f22a764a3e   
6          laurenreilly@lopez-hurst.com  60880b0bcf614ccda8688409ee9d8db3   
7  wallaceheather@obrien-richardson.com  357cbf915fdb46dfa76c7bd3ea10feff   
8                   michael18@yahoo.com  bbf4acd3660c4ba1803f396d789e538a   
9                erinsanford@garcia.com  3422bcb95ad6454f8348087ba25902f0   

                name name.firstname name.lastname name.middlename  \
0                NaN        Cynthia        Dawson          Claire   
1                NaN         Tamara         Myers             NaN   
2                NaN          Jamie     Alexander             NaN   
3                NaN         Angela        Garcia          Alexis   
4                NaN       Jennifer     Rodriguez             NaN   
5                NaN          Keith       Meadows             NaN   
6                NaN          Shane        Garcia             NaN   
7                NaN        Charles          Bell             Lee   
8      Morgan Morris            NaN           NaN             NaN   
9  Bernard Wilkerson            NaN           NaN             NaN   

                 phone          record_date          ssn  
0         624-869-4610  2006-07-08T09:02:13  xxx-xx-2412  
1       1-594-462-7759  2009-03-28T20:22:57  xxx-xx-8025  
2     472.218.5065x389  2016-08-30T20:31:39  xxx-xx-0568  
3  1-663-109-4460x1080  2001-02-15T18:50:35  xxx-xx-9825  
4         233-423-3823  2014-06-21T14:36:01  xxx-xx-9104  
5    686-305-8557x0623  2012-03-02T10:37:14  xxx-xx-4652  
6         721-135-6630  2013-01-19T12:17:21  xxx-xx-5597  
7   528-480-7488x60984  2004-03-03T07:35:59  xxx-xx-7244  
8          09316674472  2000-03-07T03:55:21  xxx-xx-7034  
9         196-454-2484  2016-11-29T21:01:43  xxx-xx-0866

## 2. Answer the following questions for each field in your list from question 1.

- What percentage of the records contain the field?
- What are the five most common values of the field?    
Note: I keep name and address field as it is if the further division is not given as shown in above data frame   


In [6]:
for col in df.columns:
    print("for column {}: ".format(col))
    print("Percentage of records in {}: {:.2f}%".format(col, (len(df[df[col].notnull()]) / len(df)) * 100))
    print("5 most common values for {}:".format(col))
    print(df[col].value_counts().sort_values(ascending=False).head(5))
    print("")

for column address: 
Percentage of records in address: 50.13%
5 most common values for address:
776 Torres Track Apt. 575\nJeffersonchester, NV 08036        1
5302 Lucas Island Apt. 946\nLake Jackieton, KS 18462-2375    1
37550 Ferrell Point\nNorth Johnland, AK 44167                1
79871 Coleman Field Suite 153\nWest Aaron, WI 84164-8891     1
389 Jacobs Highway\nCurtisburgh, IL 30266-4226               1
Name: address, dtype: int64

for column address.city: 
Percentage of records in address.city: 40.82%
5 most common values for address.city:
New Michael      61
Lake Michael     58
East Michael     57
Port Michael     56
North Michael    56
Name: address.city, dtype: int64

for column address.state: 
Percentage of records in address.state: 40.82%
5 most common values for address.state:
NC    1109
DC    1106
MI    1097
MD    1095
OH    1085
Name: address.state, dtype: int64

for column address.street: 
Percentage of records in address.street: 40.82%
5 most common values for address.st

## 3. How many distinct first names appear in this data set? Explain your procedure for identifying distinct first names        
As seen in the dataframe there are some names which are not seperated and are present in name column, also there are some prefixs like "Mr." and others listed below. so I need to eliminate them and extract the name by splitting the name in first name and last name by space. I take the two columns seperately and get the unique columns from them and then count the total number of unique values to get the required result.    
There are 1938 rows which do not have any names associated with them.     
The Number of distinct first names: 690

In [7]:
prefix = ['Mr.', 'Mrs.', 'Ms.', 'Dr.', 'Miss'] #unwanted prefix to remove

def getFirstName(name): #function to extract first name
    return [str for str in name.split(' ') if str not in prefix][0]

In [8]:
names = df[df.name.notnull()].name #use non NaN values
fname = names.apply(getFirstName)

In [9]:
fname = fname.append(df[df['name.firstname'].notnull()]['name.firstname']) #combine two columns

In [10]:
print("Number of rows having no name: {}".format(len(df[df.name.isnull() & df['name.firstname'].isnull()])))

Number of rows having no name: 1938


In [11]:
print("Number of distinct first names: {}".format(len(fname.unique())))

Number of distinct first names: 690


## 4. How many distinct street names appear in this data set? Explain your procedure for identifying distinct street names.
I created some rules based on my observation to eliminate few unwanted data in the address field and based on guidelines mentioned in https://pe.usps.com/cpim/ftp/pubs/pub28/pub28.pdf pg no 59-60 took most common street types and created a list of them    

Approach:    
1. Combine data from adrress and address.street column.     
2. remove string after Apt. and Suite     
3. remove all numeric characters as in this case there was no example of 5th street, etc    
4. Get the vocabulary of street types from USPS    
5. see if the second part of the name is in streetType and ignore all others    
6. Then calculate how many distinct street names are distinct    

In [12]:
address = df[df['address.street'].notnull()]['address.street']
address = address.append(df[df.address.notnull()].address)

In [13]:
# reference https://pe.usps.com/cpim/ftp/pubs/pub28/pub28.pdf pg no 59-60
def cleanAddress(address):
    address = address.split("\n")[0]
    address = address.split("Apt.")[0]
    address = address.split("Suite")[0]
    address = re.sub(r'[0-9]+', '', address)
    streetTypes = ['Alley','Avenue','Boulevard','Branch','Bridge','Circle','Court','Creek','Drive','Estate','Expressway','Freeway',\
                   'Highway','Motorway','Park','Pass','Parkway','Place','Road','Square','Street','Terrace','Trail']
    if address.strip().split(' ')[1] in streetTypes:
        return address.strip()


In [14]:
validNames = address.apply(cleanAddress)
validNames = validNames.dropna()
validNames.head(10)

0            David Pass
2          Young Square
12         Baker Branch
13     Bennett Motorway
49        Miller Street
77     Stephanie Avenue
102        Tammy Square
122          Potts Road
135        Curtis Creek
144        Cohen Street
dtype: object

In [15]:
print("Number of distinct street names: {}".format(len(validNames.unique())))

Number of distinct street names: 9334


## 5. What are the 5 most common US area codes in the phone number field? Explain your approach to identify the US area codes in this data set.
I use a rule based approach based on my observations of the numbers, all these rules are coded in cleanPhone() function    
Approach:    
1. I take the number and remove all number after the character 'x' (for 1-663-109-4460x1080)     
2. If the number starts from 0 and the len of the number is 11 then remove the first character which is 0 (for 09316674472)   
3. also if the number starts with 1 and it is followed by '-' then I remove the preceding 1(for 1-594-462-7759)   
4. If the number starts with '+' then I remove the first 6 characters (for +74(2)6173978109)   
5. At the end I remove all non numeric characters to get the final number.   
6. Then I take the first 3 numbers to get the area code and get the top most used ones which are 947, 996, 292, 913, 488 with their counts as displayed below   

In [16]:
def cleanPhone(number):
#     number = str(number)
    number = number.split('x')[0]
    if number[0]=='0' and len(number)==11:
        number = number[1:]
    elif number[0]=='1' and number[1]=='-':
        number = number[1:]
    if number[0]=='+':
        number = number[6:]
    number = re.sub('[^0-9]','', number)
    return number

In [17]:
phoneNos = df.phone.dropna()
areaCodes = phoneNos.apply(cleanPhone).str.slice(0,3)

In [18]:
print("Top 5 areacodes are as follows:\n{}".format(areaCodes.value_counts().sort_values(ascending=False).head(5)))

Top 5 areacodes are as follows:
947    178
996    175
292    172
913    172
488    170
Name: phone, dtype: int64
